failure to build from source with GCC6

Hello,

The Debian project is switching to GCC 6 for the next stable release. Currently bamtools fails to build with GCC6, as reported at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=811856

```

> sbuild (Debian sbuild) 0.67.0 (26 Dec 2015) on dl580gen9-02.hlinux
...
> [ 98%] Building CXX object src/toolkit/CMakeFiles/bamtools_cmd.dir/bamtools_count.cpp.o
> cd /<<BUILDDIR>>/bamtools-2.4.0+dfsg/obj-x86_64-linux-gnu/src/toolkit && /usr/bin/c++   -D_FILE_OFFSET_BITS=64 -I/<<BUILDDIR>>/bamtools-2.4.0+dfsg/src -I/<<BUILDDIR>>/bamtools-2.4.0+dfsg/obj-x86_64-linux-gnu/include -I/<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/api -I/<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/utils  -g -O2 -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2  -O3 -DNDEBUG   -Wall -o CMakeFiles/bamtools_cmd.dir/bamtools_count.cpp.o -c /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_count.cpp
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp: In member function 'bool BamTools::ResolveTool::ReadNamesFileReader::Read(std::map<std::__cxx11::basic_string<char>, ReadGroupResolver>&)':
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:413:74: error: no matching function for call to 'make_pair(__gnu_cxx::__alloc_traits<std::allocator<std::__cxx11::basic_string<char> > >::value_type&, bool)'
>          resolver.ReadNames.insert( make_pair<string,bool>(fields[1], true) ) ;
>                                                                           ^
> 
> In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
>                  from /usr/include/c++/6/bits/char_traits.h:39,
>                  from /usr/include/c++/6/string:40,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_tool.h:14,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.h:13,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:10:
> /usr/include/c++/6/bits/stl_pair.h:406:5: note: candidate: template<class _T1, class _T2> constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&)
>      make_pair(_T1&& __x, _T2&& __y)
>      ^~~~~~~~~
> 
> /usr/include/c++/6/bits/stl_pair.h:406:5: note:   template argument deduction/substitution failed:
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:413:74: note:   cannot convert 'fields.std::vector<_Tp, _Alloc>::operator[]<std::__cxx11::basic_string<char>, std::allocator<std::__cxx11::basic_string<char> > >(1ul)' (type '__gnu_cxx::__alloc_traits<std::allocator<std::__cxx11::basic_string<char> > >::value_type {aka std::__cxx11::basic_string<char>}') to type 'std::__cxx11::basic_string<char>&&'
>          resolver.ReadNames.insert( make_pair<string,bool>(fields[1], true) ) ;
>                                                                           ^
> 
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp: In member function 'bool BamTools::ResolveTool::StatsFileReader::ParseReadGroupLine(const string&, std::map<std::__cxx11::basic_string<char>, ReadGroupResolver>&)':
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:610:75: error: no matching function for call to 'make_pair(const string&, ReadGroupResolver&)'
>      readGroups.insert( make_pair<string, ReadGroupResolver>(name, resolver) );
>                                                                            ^
> 
> In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
>                  from /usr/include/c++/6/bits/char_traits.h:39,
>                  from /usr/include/c++/6/string:40,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_tool.h:14,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.h:13,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:10:
> /usr/include/c++/6/bits/stl_pair.h:406:5: note: candidate: template<class _T1, class _T2> constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&)
>      make_pair(_T1&& __x, _T2&& __y)
>      ^~~~~~~~~
> 
> /usr/include/c++/6/bits/stl_pair.h:406:5: note:   template argument deduction/substitution failed:
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:610:61: note:   cannot convert 'name' (type 'const string {aka const std::__cxx11::basic_string<char>}') to type 'std::__cxx11::basic_string<char>&&'
>      readGroups.insert( make_pair<string, ReadGroupResolver>(name, resolver) );
>                                                              ^~~~
> 
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp: In member function 'bool BamTools::ResolveTool::ResolveToolPrivate::MakeStats()':
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:1017:93: error: no matching function for call to 'make_pair(std::__cxx11::string&, const bool&)'
>          else resolver.ReadNames.insert( make_pair<string, bool>(al.Name, isCurrentMateUnique) );
>                                                                                              ^
> 
> In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
>                  from /usr/include/c++/6/bits/char_traits.h:39,
>                  from /usr/include/c++/6/string:40,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_tool.h:14,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.h:13,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:10:
> /usr/include/c++/6/bits/stl_pair.h:406:5: note: candidate: template<class _T1, class _T2> constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&)
>      make_pair(_T1&& __x, _T2&& __y)
>      ^~~~~~~~~
> 
> /usr/include/c++/6/bits/stl_pair.h:406:5: note:   template argument deduction/substitution failed:
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:1017:68: note:   cannot convert 'al.BamTools::BamAlignment::Name' (type 'std::__cxx11::string {aka std::__cxx11::basic_string<char>}') to type 'std::__cxx11::basic_string<char>&&'
>          else resolver.ReadNames.insert( make_pair<string, bool>(al.Name, isCurrentMateUnique) );
>                                                                  ~~~^~~~
> 
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp: In member function 'void BamTools::ResolveTool::ResolveToolPrivate::ParseHeader(const BamTools::SamHeader&)':
> /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:1049:93: error: no matching function for call to 'make_pair(const string&, ReadGroupResolver)'
>          m_readGroups.insert( make_pair<string, ReadGroupResolver>(rg.ID, ReadGroupResolver()) );
>                                                                                              ^
> 
> In file included from /usr/include/c++/6/bits/stl_algobase.h:64:0,
>                  from /usr/include/c++/6/bits/char_traits.h:39,
>                  from /usr/include/c++/6/string:40,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_tool.h:14,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.h:13,
>                  from /<<BUILDDIR>>/bamtools-2.4.0+dfsg/src/toolkit/bamtools_resolve.cpp:10:
> /usr/include/c++/6/bits/stl_pair.h:406:5: note: candidate: template<class _T1, class _T2> constexpr std::pair<typename std::__decay_and_strip<_Tp>::__type, typename std::__decay_and_strip<_T2>::__type> std::make_pair(_T1&&, _T2&&)
>      make_pair(_T1&& __x, _T2&& __y)
>      ^~~~~~~~~
```

Any assistance is appreciated; thanks!

--- a/src/toolkit/bamtools_resolve.cpp
+++ b/src/toolkit/bamtools_resolve.cpp
@@ -410,7 +410,7 @@ bool ResolveTool::ReadNamesFileReader::Read(map<string, ReadGroupResolver>& read
         ReadGroupResolver& resolver = (*rgIter).second;
 
         // store read name with resolver
-        resolver.ReadNames.insert( make_pair<string,bool>(fields[1], true) ) ;
+        resolver.ReadNames.insert( make_pair(fields[1], true) ) ;
     }
 
     // if here, return success
@@ -607,7 +607,7 @@ bool ResolveTool::StatsFileReader::ParseReadGroupLine(const string& line,
     resolver.IsAmbiguous = ( fields.at(6) == TRUE_KEYWORD );
 
     // store RG entry and return success
-    readGroups.insert( make_pair<string, ReadGroupResolver>(name, resolver) );
+    readGroups.insert( make_pair(name, resolver) );
     return true;
 }
 
@@ -1014,7 +1014,7 @@ bool ResolveTool::ResolveToolPrivate::MakeStats(void) {
         }
 
         // if read name not found, store new entry
-        else resolver.ReadNames.insert( make_pair<string, bool>(al.Name, isCurrentMateUnique) );
+        else resolver.ReadNames.insert( make_pair(al.Name, isCurrentMateUnique) );
     }
 
     // close files
@@ -1046,7 +1046,7 @@ void ResolveTool::ResolveToolPrivate::ParseHeader(const SamHeader& header) {
     SamReadGroupConstIterator rgEnd  = header.ReadGroups.ConstEnd();
     for ( ; rgIter != rgEnd; ++rgIter ) {
         const SamReadGroup& rg = (*rgIter);
-        m_readGroups.insert( make_pair<string, ReadGroupResolver>(rg.ID, ReadGroupResolver()) );
+        m_readGroups.insert( make_pair(rg.ID, ReadGroupResolver()) );
     }
 }
 

